jbmc/src/java_bytecode/java_trace_verification.cpp

```diff
@@ -0,0 +1,181 @@
+      continue;
+    const auto &lhs = skip_typecast(step.full_lhs);
+    const auto &rhs = skip_typecast(step.full_lhs_value);
+    if(!valid_lhs_expr_high_level(lhs))
+      print_error("LHS", lhs);
+    // check member lhs structure
+    if(const auto member = expr_try_dynamic_cast<member_exprt>(lhs))
+    {
+      if(!check_member_structure(*member))
+        print_error("LHS", lhs);
+    }
+    // check symbol lhs structure
+    if(const auto symbol = expr_try_dynamic_cast<symbol_exprt>(lhs))
+    {
+      if(!check_symbol_structure(*symbol))
+        print_error("LHS", lhs);
+    }
+    // check index lhs structure
+    if(const auto index = expr_try_dynamic_cast<index_exprt>(lhs))
+    {
+      if(index->operands().size() != 2)
+        print_error("LHS", lhs);
+      if(!check_symbol_structure(index->op0()))
+        print_error("LHS", lhs);
+      if(!expr_try_dynamic_cast<constant_exprt>(index->op1()))
+        print_error("LHS", lhs);
+    }
+    // check byte extract lhs structure
+    if(lhs.id() == ID_byte_extract_little_endian)
+    {
+      if(!check_symbol_structure(lhs.op0()))
+        print_error("LHS", lhs);
+      if(!expr_try_dynamic_cast<constant_exprt>(simplify_expr(lhs.op1(), ns)))
+        print_error("LHS", lhs);
+    }
+
+    if(!valid_rhs_expr_high_level(rhs))
+      print_error("RHS", rhs);
+    // check address_of rhs structure (String only)
+    if(const auto address = expr_try_dynamic_cast<address_of_exprt>(rhs))
+    {
+      const auto symbol_expr = get_inner_symbolexpr(*address);
+      if(!symbol_expr)
+        print_error("RHS", rhs);
+    }
+    // check symbol rhs structure (String only)
+    if(const auto symbol_expr = expr_try_dynamic_cast<symbol_exprt>(rhs))
+    {
+      if(!check_symbol_structure(*symbol_expr))
+        print_error("RHS", rhs);
+    }
+    // check struct rhs structure
+    if(const auto struct_expr = expr_try_dynamic_cast<struct_exprt>(rhs))
+    {
+      if(!struct_expr->has_operands())
+        print_error("RHS", *struct_expr);
+      if(
+        struct_expr->op0().id() != ID_struct &&
+        struct_expr->op0().id() != ID_constant)
+      {
+        print_error("RHS", *struct_expr);
+      }
+      if(
+        struct_expr->operands().size() > 1 &&
+        std::any_of(
+          ++struct_expr->operands().begin(),
+          struct_expr->operands().end(),
+          [&](const exprt &operand) { return operand.id() != ID_constant; }))
+      {
+        print_error("RHS", *struct_expr);
+      }
+    }
+    // check array rhs structure
+    if(const auto array_expr = expr_try_dynamic_cast<array_exprt>(rhs))
+    {
+      // seems no check is required.
+    }
+    // check array rhs structure
+    if(
+      const auto array_list_expr = expr_try_dynamic_cast<array_list_exprt>(rhs))
+    {
+      // seems no check is required.
+    }
+    // check constant rhs structure
+    if(const auto constant_expr = expr_try_dynamic_cast<constant_exprt>(rhs))
```

Comment:
I think there should be `else` before all these `if`s and raise an error if non of these is reached